src/firewall/server/dispatcher.py

```diff
@@ -0,0 +1,36 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+import os
+import subprocess
+
+from firewall.core.logger import log
+from firewall.config import FIREWALLD_DISPATCHER, ETC_FIREWALLD_DISPATCHER, COMMANDS
+
+
+def run_dispatcher(interface, signal, args):
+    def _dispatchers(directories):
+        for directory in directories:
+            if not os.path.isdir(directory):
+                continue
+            for file in os.listdir(directory):
+                _path = os.path.abspath(directory + os.sep + file)
+                if os.path.isfile(_path) and os.access(_path, os.X_OK):
+                    yield _path
+
+    env = {"LANG": "C"}
+    prog_args1 = [COMMANDS["systemd-run"], "--no-block", "--property", "TimeoutSec=60"]
+    prog_args2 = [interface, signal] + [str(a) for a in args]
+
+    for file in _dispatchers([FIREWALLD_DISPATCHER, ETC_FIREWALLD_DISPATCHER]):
```

Comment:
should be `list(_dispatchers(...))` to build a list of files upfront.


While `subprocess.Popen()` doesn't block, it still is maybe a bit expensive to fork and take a bit. I think we should try to build the list of scripts at once (of course, it's not atomic either, but at least do it fast).